xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/VideoCodec.h

```diff
@@ -0,0 +1,247 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#pragma once
+
+#include "../AddonBase.h"
+#include "../StreamCrypto.h"
+#include "../StreamCodec.h"
+
+#ifdef BUILD_KODI_ADDON
+#include "../DVDDemuxPacket.h"
+#else
+#include "cores/VideoPlayer/DVDDemuxers/DVDDemuxPacket.h"
+#endif
+
+namespace kodi { namespace addon { class CInstanceVideoCodec; } }
+
+extern "C"
+{
+  enum VIDEOCODEC_FORMAT
+  {
+    UnknownVideoFormat = 0,
+    VideoFormatYV12,
+    VideoFormatI420,
+    MaxVideoFormats
+  };
+
+
+  struct VIDEOCODEC_INITDATA
+  {
+    enum Codec {
+      CodecUnknown = 0,
+      CodecVp8,
+      CodecH264,
+      CodecVp9
+    } codec;
+
+    kodi::addon::CODEC_PROFILE codecProfile;
+    
+    //UnknownVideoFormat is terminator
+    VIDEOCODEC_FORMAT *videoFormats;
+
+    uint32_t width, height;
+    
+    const uint8_t *extraData;
+    unsigned int extraDataSize;
+
+    CRYPTO_INFO cryptoInfo;
+  };
+
+  struct VIDEOCODEC_PICTURE
+  {
+    enum VideoPlane {
+      YPlane = 0,
+      UPlane,
+      VPlane,
+      MaxPlanes = 3,
+    };
+
+    enum Flags :uint32_t {
+      FLAG_DROP,
+      FLAG_DRAIN
+    };
+
+    VIDEOCODEC_FORMAT videoFormat;
+    uint32_t flags;
+
+    uint32_t width, height;
+
+    uint8_t *decodedData;
+    size_t decodedDataSize;
+
+    uint32_t planeOffsets[VideoPlane::MaxPlanes];
+    uint32_t stride[VideoPlane::MaxPlanes];
+
+    int64_t pts;
+  };
+
+  enum VIDEOCODEC_RETVAL
+  {
+    VC_NONE = 0,        //< noop
+    VC_ERROR,           //< an error occured, no other messages will be returned
+    VC_BUFFER,          //< the decoder needs more data
+    VC_PICTURE,         //< the decoder got a picture
+    VC_EOF,             //< the decoder signals EOF
+  };
+
+  // this are properties given to the addon on create
+  // at this time we have no parameters for the addon
+  typedef struct AddonProps_VideoCodec
+  {
+    int dummy;
+  } AddonProps_VideoCodec;
+
+  struct AddonInstance_VideoCodec;
+  typedef struct KodiToAddonFuncTable_VideoCodec
+  {
+    kodi::addon::CInstanceVideoCodec* addonInstance;
+
+    //! \brief Opens a codec
+    bool (__cdecl* open) (const AddonInstance_VideoCodec* instance, VIDEOCODEC_INITDATA &initData);
+
+    //! \brief Reconfigures a codec
+    bool (__cdecl* reconfigure) (const AddonInstance_VideoCodec* instance, VIDEOCODEC_INITDATA &initData);
+
+    //! \brief Feed codec if requested from GetPicture() (return VC_BUFFER)
+    bool (__cdecl* add_data) (const AddonInstance_VideoCodec* instance, const DemuxPacket &packet);
+
+    //! \brief Get a decoded picture / request new data
+    VIDEOCODEC_RETVAL (__cdecl* get_picture) (const AddonInstance_VideoCodec* instance, VIDEOCODEC_PICTURE &picture);
```

Comment:
Yep, must be plain C. No references in C, they are a C++ language construct.

Having a plain C api gateway has the advantage that this is the base  for adding support for other add-on programming  languages, because most programming languages can call C (e.g. Java using JNI), but no C++.